### PR TITLE
[Layer] Add trainable prop

### PR DESCRIPTION
**Pull request by zhoonit**
- [Layer] Add trainable prop

```
This patch add trainable property to layer

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: Jihoon Lee <jhoon.it.lee@samsung.com>

```

### CODE REVIEW

File: nntrainer/layers/layer_impl.cpp [hunk 1]
```diff
@@ -23,7 +24,7 @@ namespace nntrainer {
 
 LayerImpl::LayerImpl() :
   finalized(false),
-  layer_impl_props(std::make_unique<std::tuple<>>()) {}
```
Comment (kparichay) [inline nntrainer/layers/layer_impl.cpp]: props trainable will be in LayerNode and not in LayerImpl, and will be accessible to Layer via Init/Run context.

Comment (zhoonit) [reply nntrainer/layers/layer_impl.cpp]: Thought about it, there are some issue

1. What if a layer is not trainable by default, or must be always non trainable?
2. When we request tensor, only layer knows which layer should be trainable or non-trainable
    eg) Even if we set trainable, weight1 and weight2 is trainable but weight3, weight4 must stay non-trainble
          Even if we set non-trainable, weight1 must be trainable.

Comment (zhoonit) [reply nntrainer/layers/layer_impl.cpp]: My thinking is that layer has the boolean trainable but when request weight set true or false respectively.

Comment (kparichay) [reply nntrainer/layers/layer_impl.cpp]: > Thought about it, there are some issue
> 
> 1. What if a layer is not trainable by default, or must be always non trainable?

yeah, there are pros and cons. To support `always non-trainable`, there will yet another function in layer devel like `supportTraining()` if trainable in LayerNode. If in LayerImpl, we need `getTrainable` or `getProperty` now. Both fine choices, left up to you.

> 2. When we request tensor, only layer knows which layer should be trainable or non-trainable
>    eg) Even if we set trainable, weight1 and weight2 is trainable but weight3, weight4 must stay non-trainble
>    Even if we set non-trainable, weight1 must be trainable.

yes, that must be controlled by the layer. When layer requestWeights() by context, layer gets full control of this. This true/false for each weight must be `and` with the trainable of the layer, and then correspondingly set to weight (not yet done, just my thought).

Comment (kparichay) [reply nntrainer/layers/layer_impl.cpp]: If you choose to keep trainable in LayerImpl, please update LayerNode correspondingly as well (remove trainable from there).

Comment (zhoonit) [reply nntrainer/layers/layer_impl.cpp]: Thanks for the suggestion, I'll update within this PR.

Could you also take a look over #1297 regarding this?

Comment (kparichay) [reply nntrainer/layers/layer_impl.cpp]: done.

Comment (zhoonit) [reply nntrainer/layers/layer_impl.cpp:26]: > I'll update within this PR.

trainable deleted from props

**Review by kparichay — APPROVED** [review]
LGTM

File: nntrainer/layers/layer_impl.cpp (hunk 1 above)
Comment (jijoongmoon) [reply nntrainer/layers/layer_impl.cpp:26]: >When we request tensor, only layer knows which layer should be trainable or non-trainable
eg) Even if we set trainable, weight1 and weight2 is trainable but weight3, weight4 must stay non-trainble
Even if we set non-trainable, weight1 must be trainable.

I wonder whether we are going to set the trainable in weight ??

**Comment by kparichay** [comment]
Discussed offline, trainable will later move to LayerNode.
Lets merge this PR for now.

File: nntrainer/layers/layer_impl.cpp (hunk 1 above)
Comment (zhoonit) [reply nntrainer/layers/layer_impl.cpp:26]: Yes, we are setting trainable in weights and that won't be changing.

**Review by jijoongmoon — APPROVED** [review]
LGTM

**Pull request merged**